### PR TITLE
feat: add formTemplate model and api

### PR DESCRIPTION
**Pull request by uyousafzai54**
## Notion ticket link

[Ticket Name](https://www.notion.so/uwblueprintexecs/Tasks-38b49f83f6ce4b07ae188dabb37631e8?p=be8256776a614986a6d3df8f9b9fbabf)

## Implementation description
* Created a new Form Template model that contains an array of FormQuestion Object IDs.
* Created two new endpoints in the admin service for creating/updating and getting the Form Template

## Steps to test
1. Hit the following endpoint `http://localhost:5000/admin/formTemplate` with a POST request containing the following body below.
```
{
  "formQuestions": [
    {
      "question": "q1",
      "type": "Multiselect",
      "required": true,
      "description": "question description"
    },
    {
      "question": "q2",
      "type": "Multiselect",
      "required": false,
      "description": "question description",
      "options": [
        "'option1'"
      ]
    }
  ]
}
```
2. Hit the following endpoint `http://localhost:5000/admin/formTemplate` with a GET request to get the resulting array. 

## What should reviewers focus on?
* API Implementation. 

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: backend/typescript/services/interfaces/adminService.ts [hunk 1]
```diff
@@ -14,6 +14,18 @@ interface IAdminService {
    * @throws Error if waiver retrieval fails.
    */
   getWaiver(): Promise<WaiverDTO>;
+  /**
+   * Creates a new form template or updates it if it already exists.
+   * @returns A WaiverDTO that contains the waiver.
```
Comment (amanda-guo) [inline backend/typescript/services/interfaces/adminService.ts:19]: nit: update this @returns to a FormTemplateDTO

File: backend/typescript/services/interfaces/adminService.ts [hunk 2]
```diff
@@ -14,6 +14,18 @@ interface IAdminService {
...
+   * @throws Error if waiver retrieval fails.
```
Comment (amanda-guo) [inline backend/typescript/services/interfaces/adminService.ts:20]: nit: update this @throws to be specific to a form template

File: backend/typescript/services/interfaces/adminService.ts [hunk 3]
```diff
@@ -14,6 +14,18 @@ interface IAdminService {
...
+   */
+  updateFormTemplate(form: FormTemplateDTO): Promise<FormTemplateDTO>;
+  /**
+   * Get the form.
+   * @returns A WaiverDTO that contains the waiver.
```
Comment (amanda-guo) [inline backend/typescript/services/interfaces/adminService.ts:25]: nit: update this @returns to a FormTemplateDTO

File: backend/typescript/services/interfaces/adminService.ts [hunk 4]
```diff
@@ -14,6 +14,18 @@ interface IAdminService {
...
+   * @throws Error if waiver retrieval fails.
```
Comment (amanda-guo) [inline backend/typescript/services/interfaces/adminService.ts:26]: nit: update this @throws to be specific to a form template

**Review by amanda-guo — COMMENTED** [review]
Is there only supposed to be 1 formTemplate that can exist in the whole system? Since that's what's implemented right now. But according to the architecture doc, it seems the formTemplates are camp-specific?

File: backend/typescript/services/implementations/adminService.ts [hunk 5]
```diff
@@ -47,6 +49,61 @@ class AdminService implements IAdminService {
...
+        },
+        {
+          $set: { formQuestions },
+        },
+        { upsert: true, runValidators: true },
+      );
+      formTemplateDTO = await this.getFormTemplate();
+    } catch (error: unknown) {
+      Logger.error(
+        `Failed to update waiver. Reason: ${getErrorMessage(error)}`,
```
Comment (amanda-guo) [inline backend/typescript/services/implementations/adminService.ts:82]: nit: update error message to be specific to form templates

File: backend/typescript/services/implementations/adminService.ts [hunk 6]
```diff
@@ -47,6 +49,61 @@ class AdminService implements IAdminService {
...
+      form = await formTemplateModel.findOne();
+      console.log(form?.formQuestions);
+      if (!form) {
+        throw new Error(`Form not found.`);
+      }
+      formTemplateDTO = {
+        formQuestions: form.formQuestions,
+      };
+    } catch (error: unknown) {
+      Logger.error(`Failed to get waiver. Reason = ${getErrorMessage(error)}`);
```
Comment (amanda-guo) [inline backend/typescript/services/implementations/adminService.ts:102]: nit: update error message to be specific to form templates

File: backend/typescript/services/implementations/adminService.ts [hunk 7]
```diff
@@ -47,6 +49,61 @@ class AdminService implements IAdminService {
...
+    }
+    return formTemplateDTO;
+  }
+
+  async getFormTemplate(): Promise<FormTemplateDTO> {
+    let formTemplateDTO: FormTemplateDTO | null;
+    let form: FormTemplateDTO | null;
+    try {
+      form = await formTemplateModel.findOne();
+      console.log(form?.formQuestions);
```
Comment (amanda-guo) [inline backend/typescript/services/implementations/adminService.ts:94]: nit: can remote this console.log

**Review by amanda-guo — COMMENTED** [review]
another thing - it's mentioned in the notion ticket "Get/edit the questions in the template." Was that also implemented in the scope of this PR?

File: backend/typescript/services/implementations/userService.ts [hunk 8]
```diff
@@ -253,7 +253,7 @@ class UserService implements IUserService {
                   lastName: oldUser.lastName,
                   role: oldUser.role,
                   active: oldUser.active,
-                  camps: (oldUser as CampLeader).camps,
+                  camps: (oldUser as CampLeader).campSessions,
```
Comment (annechung) [inline backend/typescript/services/implementations/userService.ts:256]: think we need to change all the `camps` in this file to `campSessions`  - do you mind making that change for now? @amanda-guo should be updating some of this code in her PR as well

File: backend/typescript/services/implementations/adminService.ts [hunk 9]
```diff
@@ -47,6 +49,61 @@ class AdminService implements IAdminService {
...
+      throw error;
+    }
+    return formTemplateDTO;
+  }
+
+  async getFormTemplate(): Promise<FormTemplateDTO> {
+    let formTemplateDTO: FormTemplateDTO | null;
+    let form: FormTemplateDTO | null;
+    try {
+      form = await formTemplateModel.findOne();
```
Comment (annechung) [inline backend/typescript/services/implementations/adminService.ts:93]: you need to populate the formQuestions here so we return the full formQuestion objects and not just the ids; an example of populating here: https://github.com/uwblueprint/focus-on-nature/blob/main/backend/typescript/services/implementations/camperService.ts#L150

**Review by annechung — CHANGES_REQUESTED** [review]
good start! just need to make some fixes to return the full formquestion objects (in both get and update handlers) and to remove waiver things :')

File: backend/typescript/services/implementations/adminService.ts [hunk 10]
```diff
@@ -47,6 +51,65 @@ class AdminService implements IAdminService {
     }
     return waiverDto;
   }
+
+  async updateFormTemplate(form: FormTemplateDTO): Promise<FormTemplateDTO> {
+    let formTemplateDTO: FormTemplateDTO | null;
+    try {
+      /* eslint no-underscore-dangle: 0 */
```
Comment (annechung) [inline backend/typescript/services/implementations/adminService.ts:58]: we should delete the existing form questions before creating new ones

File: backend/typescript/services/implementations/adminService.ts [hunk 11]
```diff
@@ -47,6 +51,58 @@ class AdminService implements IAdminService {
...
+            type: formQuestion.type,
+            question: formQuestion.question,
+            required: formQuestion.required,
+            description: formQuestion.description,
+            options: formQuestion.options,
+          });
+          formQuestions[i] = question._id;
+        }),
+      );
+      await formTemplateModel.remove({});
```
Comment (annechung) [inline backend/typescript/services/implementations/adminService.ts:72]: ah i think we want to delete the questions in the actual formQuestions collection? so old unused questions aren't hanging around. so before creating the new question documents we should delete all current questions in formQuestions

**Review by AmandaMisjuwar — APPROVED** [review]

**Review by annechung — APPROVED** [review]

**Pull request merged**